videostore/mux.go

```diff
@@ -0,0 +1,573 @@
+	}
+	spsUnChanged       bool
+	firstTimeStampsSet bool
+	firstPTS           int64
+	firstDTS           int64
+}
+type rawSegmenterMux struct {
+	// are valid for the lifetime of the rawSegmenterMux
+	camName resource.Name
+	logger  logging.Logger
+	worker  *utils.StoppableWorkers
+	regDone <-chan struct{}
+	cam     registry.ModuleCamera
+
+	mu       sync.Mutex
+	rawSeg   *videostore.RawSegmenter
+	codec    atomic.Int64
+	metadata metadata
+}
+
+var codecs = []videostore.CodecType{
+	videostore.CodecTypeH265,
+	videostore.CodecTypeH264,
+}
+
+func newRawSegmenterMux(rawSeg *videostore.RawSegmenter, camName resource.Name, logger logging.Logger) *rawSegmenterMux {
+	return &rawSegmenterMux{
+		rawSeg:  rawSeg,
+		camName: camName,
+		worker:  utils.NewBackgroundStoppableWorkers(),
+		logger:  logger,
+	}
+}
+
+// init and close are called by videostore.
+func (m *rawSegmenterMux) init() error {
+	if m.rawSeg == nil {
+		return errors.New("videostore.RTPVideoStore.Segmenter() is nil")
+	}
+	cam, err := registry.Global.Get(m.camName.String())
+	if err != nil {
+		return fmt.Errorf("resource %s unable to be found in viamrtsp registry err: %s", m.camName.String(), err.Error())
+	}
+	regCtx, err := cam.RequestVideo(m, codecs)
+	if err != nil {
+		return err
+	}
+	m.regDone = regCtx.Done()
+	m.cam = cam
+	m.worker.Add(m.registrationMonitor)
+	return nil
+}
+
+func (m *rawSegmenterMux) registrationMonitor(ctx context.Context) {
+	registered := true
+	timer := time.NewTimer(monitorInterval)
+	defer timer.Stop()
+	defer m.cleanup()
+	for {
+		if err := ctx.Err(); err != nil {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-m.regDone:
+			registered = false
+			// set the channel to nil so we don't go down this case statement next iteration
+			m.regDone = nil
+
+		case <-timer.C:
+			if !registered {
+				cam, err := registry.Global.Get(m.camName.String())
+				if err != nil {
+					m.logger.Warnf("failed to find camera %s", err.Error())
+					timer.Reset(monitorInterval)
+					continue
+				}
+
+				regCtx, err := cam.RequestVideo(m, codecs)
+				if err != nil {
+					m.logger.Warnf("failed to register video-store with viamrtsp camera, err: %s", err.Error())
+					timer.Reset(monitorInterval)
+					continue
+				}
+				m.regDone = regCtx.Done()
+				m.cam = cam
+				registered = true
+				timer.Reset(monitorInterval)
+				continue
+			}
+
+			if videostore.CodecType(m.codec.Load()) == videostore.CodecTypeUnknown {
+				m.logger.Warn("waiting for viamrtsp camera to send video data to video-store")
+			}
+			timer.Reset(monitorInterval)
+		}
+	}
+}
+
+func (m *rawSegmenterMux) cleanup() {
+	if err := m.Stop(); err != nil {
+		m.logger.Warnf("failed to stop raw segmenter %s", err.Error())
+	}
+	if err := m.cam.CancelRequest(m); err != nil {
+		m.logger.Warnf("DeRegister video-store from viamrtsp camera %s", err.Error())
+	}
+}
+
+func (m *rawSegmenterMux) close() error {
+	if m == nil {
+		return nil
+	}
+	m.worker.Stop()
```

Comment:
I don't know how we'd get into this state in this particular - but usually we also check that the worker is nil before stopping it.